stdlib/3/concurrent/futures/_base.pyi

```diff
@@ -31,12 +33,23 @@ class Future(Generic[_T]):
     def set_result(self, result: _T) -> None: ...
     def set_exception(self, exception: Optional[BaseException]) -> None: ...
 
-class Executor:
-    def submit(self, fn: Callable[..., _T], *args: Any, **kwargs: Any) -> Future[_T]: ...
-    def map(self, func: Callable[..., _T], *iterables: Iterable[Any], timeout: Optional[float] = ..., chunksize: int = ...) -> Iterator[_T]: ...
-    def shutdown(self, wait: bool = ...) -> None: ...
-    def __enter__(self) -> Executor: ...
```

Comment:
Why can't you achieve what you want with the simpler solution of typing `__enter__` as `def __enter__(self: _T) -> _T: ...`, where `_T` is a TypeVar bound to Executor? That would make the rest of this PR unnecessary and remove the version_info checks.

Reply:
That would work, too. 

I didn't do it that way because of the note [here](https://mypy.readthedocs.io/en/latest/generics.html#generic-methods-and-generic-self) about generic selfs being experimental and not fully implemented.  

But if you think that's the way forward I can rework it like that.

Reply:
But self is not generic here, is it? Executor itself is not a generic class.

Reply:
I just meant "generic self"  to mean that `self` has a TypeVar type (which I believe is what the note is saying).

According to the note, I don't think it matters if the class (Executor) that gets bound to the TypeVar is itself a 'generic class'.

See the [example](https://mypy.readthedocs.io/en/latest/generics.html#generic-methods-and-generic-self), none of the `Shape`, `Circle`, or `Square` classes are generic either.

Reply:
My understanding is that self-types are fairly stable for non-generic classes (like the ones we're using here). The troubles arise when the class we're using a self type for is itself generic: see python/mypy#2354, python/mypy#3363, python/mypy#3631.

Reply:
In any case, self types are already widely used in typeshed for non-generic classes (grep finds about 130 usages). I'm confident that using them here too is safe.

Reply:
Okay, then!  I'll rework this that way when I get a moment.